### PR TITLE
ath79: add support for Netgear R6100

### PR DESCRIPTION
**Pull request by wigyori**
Netgear R6100 AC1200 dual-band wifi router.

Specifications:

SoC:    QCA9344
DRAM:	128Mb DDR2
Flash:	128Mb NAND flash
Switch:	5x 10/100 via AR8229 switch integrated into SoC
WLAN:	AR9344 (SoC, 2.4G, 2x2) + QCA9882 (PCIe, 5G, 2x2)
USB:	1x 2.0
UART:	yes
LEDs:	4 + 2
Button:	3 (reset, WPS, wifi on/off)

Installation:

The factory image can be uploaded via the original web interface, which
is the easiest method to replace it. For subsequent upgrades, please use
the standard sysupgrade method.

Alternative installation / recovery:

The factory image can be uploaded with nmrpflash ("unofficial flash tool for
Netgear network devices") at boot-time. U-boot waits for a couple seconds
for a Netgear-compatible image as part of the boot process before loading
the kernel.

MAC addresses as verified by OEM firmware:
LAN	*:5D	caldata 0x0
WAN	*:5E	caldata 0x6
WLAN2G	*:5D	caldata 0x0
WLAN5G	*:5F	caldata 0x6 + 1

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

### CODE REVIEW

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (adschm) [inline target/linux/ath79/dts/ar9344_netgear_r6100.dts:47]: "leds:" DT label can be dropped.

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (adschm) [inline target/linux/ath79/dts/ar9344_netgear_r6100.dts:64]: wlan_blue or just wlan

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (adschm) [inline target/linux/ath79/dts/ar9344_netgear_r6100.dts:70]: usb_blue or just usb

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (adschm) [inline target/linux/ath79/dts/ar9344_netgear_r6100.dts:90]: "gpio_usb_power" label can be dropped.

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (adschm) [inline target/linux/ath79/dts/ar9344_netgear_r6100.dts:162]: Do you really need this firmware partition here?

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (adschm) [inline target/linux/ath79/dts/ar9344_netgear_r6100.dts:144]: partition@ for these two as well?

File: target/linux/ath79/nand/base-files/etc/board.d/01_leds
Comment (adschm) [inline target/linux/ath79/nand/base-files/etc/board.d/01_leds:15]: If it's a WAN LED, we should not hijack it for LAN in default setup...

File: target/linux/ath79/generic/base-files/etc/board.d/02_network
Comment (adschm) [inline target/linux/ath79/generic/base-files/etc/board.d/02_network:352]: This should be dropped, we are on nand subtarget.

File: target/linux/ath79/nand/base-files/etc/board.d/02_network
Comment (adschm) [inline target/linux/ath79/nand/base-files/etc/board.d/02_network:24]: LAN will be set automatically by the switch command, so 
ucidef_set_interface_wan "eth1"
will be enough for the first line.

File: target/linux/ath79/nand/base-files/etc/board.d/02_network
Comment (adschm) [inline target/linux/ath79/nand/base-files/etc/board.d/02_network:26]: The suffixes can be dropped (only needed if they deviate from the internal ports):
"0@eth0" "1:lan" "2:lan" "3:lan" "4:lan"

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
Comment (adschm) [inline target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:15]: No (valid) MAC address in 0x1002?

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
Comment (adschm) [inline target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:21]: No valid MAC address in 0x5006?

**Review by adschm — CHANGES_REQUESTED** [review]
Just had a quick first look ... Nothing complicated...

**Comment by mpratt14** [comment]
@adschm ironically I was also about to post a PR for this device....should I still be working on it?

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (wigyori) [reply target/linux/ath79/dts/ar9344_netgear_r6100.dts:144]: ar9344_netgear_wndr.dtsi has ubi@ and kernel@ in the DTS too, not partition@ - I thought about following that, but can change if required.

**Comment by mpratt14** [comment]
also it is worth mentioning, this DTS is partly based on one for similar Netgear devices, which also has mistakes in partitioning and other minor issues

target/linux/ath79/dts/ar9344_netgear_wndr.dtsi

File: target/linux/ath79/nand/base-files/etc/board.d/01_leds
Comment (wigyori) [reply target/linux/ath79/nand/base-files/etc/board.d/01_leds:15]: The WAN LED is actually two LEDs - one green, one amber -, it's again the WNDRs which are set up similarly:

```
netgear,wndr4500-v3)
        ucidef_set_led_switch "wan-amber" "WAN (amber)" "amber:wan" "switch0" "0x20"
```

File: target/linux/ath79/generic/base-files/etc/board.d/02_network
Comment (wigyori) [reply target/linux/ath79/generic/base-files/etc/board.d/02_network:352]: Indeed, that's a stray change.

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
Comment (wigyori) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:21]: 0x5006 contains a valid, but dummy MAC (00:03:07:12:34:56). To match the original firmware's MACs, the MAC from caldata should be taken as a base. (Actually, caldata 0x12 contains a proper MAC for the 5G interface, so we could even use that.)

**Comment by wigyori** [comment]
> also it is worth mentioning, this DTS is partly based on one for similar Netgear devices, which also has mistakes in partitioning and other minor issues
> 
> target/linux/ath79/dts/ar9344_netgear_wndr.dtsi

That assumption is fully correct that I've based my DTS on similar ones (which supposedly weren't fixed either).

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
Comment (wigyori) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:15]: There is, but I guess we still need to extract the caldata contents.

**Comment by pepe2k** [comment]
@wigyori @adschm  I have support for this one already in my staging tree: https://git.openwrt.org/openwrt/staging/pepe2k.git

**Comment by wigyori** [comment]
> @wigyori @adschm I have support for this one already in my staging tree: https://git.openwrt.org/openwrt/staging/pepe2k.git

*sigh* Whatever you prefer then :D

-w-

**Comment by pepe2k** [comment]
@wigyori let me compare our changsets and if there are differences, I will ask for your opinion. Did you runtime test your version?

**Comment by wigyori** [comment]
> @wigyori let me compare our changsets and if there are differences, I will ask for your opinion. Did you runtime test your version?

Yep, it's running on my desk. I'll have to test the changes with @adschm suggested.

File: target/linux/ath79/dts/ar9344_netgear_r6100.dts
Comment (adschm) [reply target/linux/ath79/dts/ar9344_netgear_r6100.dts]: The general recommendation is to use generic names there, i.e. just partition@ for all partitions in this case. Of course, you need to have the label property set then. If something different than partition@ is used, it's either old or slipped through review AFAIK.

File: target/linux/ath79/nand/base-files/etc/board.d/01_leds
Comment (adschm) [reply target/linux/ath79/nand/base-files/etc/board.d/01_leds:15]: But 0x20 matches (only) the WAN LED for the wndr4500-v3, and it has nothing set for green. So, you have one LED named "WAN" that indicates WAN port status there.

In your case, you have two LEDs named WAN of which one shows the WAN port status and the other one shows the LAN ports' status. I'd prefer that for default setup you choose one of the colors and link that to eth1. Some Netgear devices also have a setup where one color is for link status and the other one is for activity; however, I'd make that dependent on what the vendor did.

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
Comment (adschm) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:15]: Yes, but then you don't need to overwrite it with nvmem-cells in DTS (at &wmac).

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
Comment (adschm) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:21]: But you are using 0xc at the moment in DTS?

**Comment by adschm** [comment]
@pepe2k From a quick look at your version I see:
* different label-mac-device
* gpio-export DT label needs to be dropped as well
* swapped MAC address for eth0/eth1 (which explains the label-mac-device above)
* mtd-cal-data AND setup in 10-ath9k-eeprom
* different port assignment for LuCI
* different setup for ath10k MAC address
* different DEVICE_PACKAGES, haven't checked which makes more sense

"Different" does not imply any judgement, since I cannot (and did not try to) judge who is right here.

But since there is three people with that device here now, there shouldn't be a problem to sort everything out.

**Comment by pepe2k** [comment]
Hi @adschm, thanks for checks!

> different label-mac-device

Double checked and fixed in my version - the vendor uses LAN/WLAN 2.4 GHz MAC address on the sticker.

I suppose all of that network configuration related differences come from the fact that this device in ar71xx had WAN on eth0 (see: https://github.com/openwrt/openwrt/blob/openwrt-19.07/target/linux/ar71xx/base-files/etc/board.d/02_network#L357). The ifname name assignment (GMACx vs. ethX) has changed during transition to ath79 and now we have a lot of mess with that... personally, I always preferred to keep LAN on eth0 and WAN on eth1 (I suppose it was or maybe still is, kind of an unwritten rule for no-DSA targets) but I'm also fine with different approach for this device as there is no way to upgrade from previous release due to kernel partition size change (it got bigger and the kernel is kept outside UBI in a RAW partition as the U-Boot doesn't understand UBI).

> gpio-export DT label needs to be dropped as well

Good catch, dropped, thanks!

> swapped MAC address for eth0/eth1 (which explains the label-mac-device above)

As above.

> mtd-cal-data AND setup in 10-ath9k-eeprom

We still carry this patch: https://github.com/openwrt/openwrt/blob/master/package/kernel/mac80211/patches/ath9k/552-ath9k-ahb_of.patch but to be honest, I'm not that sure what approach is preferred now? The `mtd-cal-data` is enough here and the ART also includes proper MAC address for the ath9k radio. No idea why other similar devices use userspace method though.

> different port assignment for LuCI

Mine were confirmed with LuCi (port label on the device matches name in LuCi switch configuration).

> different setup for ath10k MAC address

DTS only based version from @wigyori looks better, I updated my patch.

> different DEVICE_PACKAGES, haven't checked which makes more sense

The ath9k is already built-in for this target and `kmod-usb-ledtrig-usbport` is included in `$(Device/netgear_ath79_nand)`

I updated my version with above changes. There are only few things left for discussion:

- DTS: `firmware` partition: I removed it from my patch, we don't need it (maybe someone though it could be used for OpenWrt -> vendor firmware change)
- DTS: `enable_gpio_11`: didn't need it on my device, the U-Boot already set up all the GPIOs for LEDs
- power LED: I followed previous support and vendor's firmware (green for `booting` and `running`)
- WAN LED: same as above, the vendor keeps amber WAN LED on when there is no Internet access (switches to green otherwise)
- WLAN LED: I use it for 2.4 GHz radio, @wigyori seems to be using it for 5 GHz, vendor uses this LED for both devices at the same time
- DTS: `gpio-export` instead of `gpio-hog`: that's my personal preference which allows end user to (easily) control USB power
- DTS: `chosen`: already defined in ar934x.dtsi
- @adschm should we also add `DEVICE_COMPAT_VERSION` and `DEVICE_COMPAT_MESSAGE` for the kernel partition size change warning? Sysupgrade from <= 19.07 will fail anyway (the kernel partition size and supported devices don't match) but I haven't actually followed discussion about the COMPAT_VERSION and don't know the correct approach here.

@wigyori any comments? Are you fine with adding your SoB to my version and push it upstream?

Cheers,
Piotr

**Comment by mpratt14** [comment]
@pepe2k Can we wait until I am done with my version? some things I want to address...

 - read-only for all OEM partitions?
 - using both green and amber for WAN led (link / rx)
 - eth1 should have MAC from 0x6, eth0 should have MAC from 0x0 (OEM has them flipped) (keep label-mac-device as eth0)
 - compatible for ath10k is `qcom,ath10k`
 - lets include flash model in commit message MT29F1G08ABAEAWP
 - (this is very picky and obsessive but) lets get used to node name with 3 digits representing address size `wifi@0,0,0`

Also, I flashed from OEM webpage and got hard brick, my DTS is a little different right now but this error cannot be attributed to DTS. I'm going to try again though

```
Loading from device 0: ath-nand (offset 0x1a0000)

** check kernel image **
   Verifying Checksum ... OK

** check rootfs image **

** Bad Magic Number 0x27051956 **

The Router is in TFTP Server Firmware Recovery mode NOW!
Listening on Port : 69, IP Address: 192.168.1.1...

```

**Comment by pepe2k** [comment]
@mpratt14 
> @pepe2k Can we wait until I am done with my version? some things I want to address...

OK.

> read-only for all OEM partitions?

No opinion on my side, I don't have much experience with Netgear stuff, I have no idea what those partition contain and are used for in vendor's firmware.

> using both green and amber for WAN led (link / rx)

In a generic support I prefer simplicity (and follow vendor's approach).
It's up to the user how the "extra" WAN LED will be used.

> eth1 should have MAC from 0x6, eth0 should have MAC from 0x0 (OEM has them flipped) (keep label-mac-device as eth0)

I think you miss the point here. ethX and ethY are only kernel's names for GMAC0 and GMAC1, you can assign names however you want and even more, you can swap GMACs, so there isn't anything like "eth0 should have MAC from 0x0" (maybe you referred to GMAC0?). I prefer to look at this like "WAN port should use the same MAC address as in vendor's firmware".

> compatible for ath10k is `qcom,ath10k`

OK.

> lets include flash model in commit message MT29F1G08ABAEAWP

Can you confirm all manufactured device have that one? The one I was given has MT29F1G08ABADA... if you really want to include information about flash, stick to the family/series name only (MT29F1G08).

> (this is very picky and obsessive but) lets get used to node name with 3 digits representing address size `wifi@0,0,0`

No opinion.

> Also, I flashed from OEM webpage and got hard brick, my DTS is a little different right now but this error cannot be attributed to DTS. I'm going to try again though

Check the image recipe for these devices, their images require fake rootfs:
`# fake rootfs is mandatory, pad-offset 129 equals (2 * uimage_header + 0xff)`

I did all my testing on the latest available vendor firmware: `R6100-V1.0.1.28`

Cheers,
Piotr

**Comment by adschm** [comment]
> > mtd-cal-data AND setup in 10-ath9k-eeprom
> 
> We still carry this patch: https://github.com/openwrt/openwrt/blob/master/package/kernel/mac80211/patches/ath9k/552-ath9k-ahb_of.patch but to be honest, I'm not that sure what approach is preferred now? The `mtd-cal-data` is enough here and the ART also includes proper MAC address for the ath9k radio. No idea why other similar devices use userspace method though.

I've never understood this properly. My uneducated impression from other people's comments was that it's _not_ same, but only one of them is right in a given situation. But this could be completely wrong either. I'd be happy if somebody could actually resolve this question, there is at least one other PR waiting for an answer as well.

**Comment by adschm** [comment]
> I suppose all of that network configuration related differences come from the fact that this device in ar71xx had WAN on eth0 (see: https://github.com/openwrt/openwrt/blob/openwrt-19.07/target/linux/ar71xx/base-files/etc/board.d/02_network#L357). The ifname name assignment (GMACx vs. ethX) has changed during transition to ath79 and now we have a lot of mess with that... personally, I always preferred to keep LAN on eth0 and WAN on eth1 (I suppose it was or maybe still is, kind of an unwritten rule for no-DSA targets) but I'm also fine with different approach for this device as there is no way to upgrade from previous release due to kernel partition size change (it got bigger and the kernel is kept outside UBI in a RAW partition as the U-Boot doesn't understand UBI).

I tend to not give ar71xx too much weight here, but try to make it nice and proper with ath79 now (if that's possible and not arbitrary ....)

It's probably easiest for most developers if GMAC and eth numbers align, i.e. &eth0 in DTS corresponds to eth0 at runtime. If I was asked, I'd set it up this way; I personally don't care much whether LAN/switch is eth0 or eth1 then. But having eth0 correspond to &eth1 is not really helpful....

And yes, there is only one way to properly assign the MAC addresses then, you just have to do the right number of swaps in your head :-)

**Comment by mpratt14** [comment]
@pepe2k 

> I think you miss the point here. ethX and ethY are only kernel's names for GMAC0 and GMAC1

ok more specifically, I mean OEM has eth0 and eth1 flipped for each GMAC compared to ath79:

for OEM (and ar71xx probably)
```
GMAC0 (WAN) -----> eth0 --> MAC at 0x6
GMAC1 (switch) --> eth1 --> MAC at 0x0
```

and in ath79
```
GMAC0 (WAN) -----> eth1 --> MAC at 0x6
GMAC1 (switch) --> eth0 --> MAC at 0x0
```
in other words, the LAN internal switch shares MAC with 2.4 GHz WLAN
(which also makes more sense than WLAN sharing a MAC with the WAN port)
and the MAC at 0x0 is the one on the label

plugging WAN port on OEM results in
```
enet0 port0 up 100Mbps Full duplex
ATHR_GMAC: Enet Unit:0 PHY:0 is UP Mii 100Mbps full duplex
ATHR_GMAC: done cfg2 0x7115 ifctl 0x10000 miictrl
Setting Drop CRC Errors, Pause Frames and Length Error frames
br1: port 1(eth0) entering forwarding state
Performing a DHCP renew

*********************
have done something about plug in
***********************
Sending discover...
enet0 port0 down
ATHR_GMAC:unit 0: phy 0 not up carrier 1
br1: port 1(eth0) entering disabled state

*********************
have done something about plug off
***********************
```

and ifconfig on OEM shows:
```
eth0      Link encap:Ethernet  HWaddr A0:63:91:88:83:13
```

on my device:
```
MAC at 0x0 --- :12
MAC at 0x6 --- :13
MAC at 0xc --- :14
```



> if you really want to include information about flash, stick to the family/series name only (MT29F1G08).

I agree 👍🏼 



> Check the image recipe for these devices,

I realized I made a silly mistake in the makefile, working too fast lol

**Comment by mpratt14** [comment]
@adschm 

> It's probably easiest for most developers if GMAC and eth numbers align, i.e. &eth0 in DTS corresponds to eth0 at runtime.

from what I see right now in ath79: the GMAC and ethX numbers do NOT align, but the ethX numbers between DTS and runtime DO align. At least this is how it is for the current device we are talking about, might be different for some.
(the only way I can tell is from using DTS to assign MAC address)

Another way to look at ethX number assignment consistently, is that some devices are switch-only, so GMAC0 is not used at all. So why would we have a device that has eth1 but no eth0? If I understand your position correctly, agreement between DTS and runtime device numbers take priority, and I agree with that.




> > to be honest, I'm not that sure what approach is preferred now? The `mtd-cal-data` is enough here and the ART also includes proper MAC address for the ath9k radio. No idea why other similar devices use userspace method though.
> 
> I'd be happy if somebody could actually resolve this question, there is at least one other PR waiting for an answer as well.

I was also curious so I did some searching and found the following mail:
http://lists.openwrt.org/pipermail/openwrt-devel/2019-August/024206.html

"mtd-cal-data is a local hack for wmac devices only"

from looking at the driver a little closely, see:
target/linux/generic/files/include/linux/ath9k_platform.h

note that eeprom_data is instantiated with a static length that is 0x1000 (0x800 number of 16-bit words)

if the DTS property is used for all devices, this could maybe cause a concern where some other data is copied that exists after the cal-data ends but is within 0x1000 from start of cal-data

So the real question at the end of the day seems to be this:

Do we prefer a local DTS hack over userspace script?

And if we do prefer the DTS hack, it would need to be updated to accomodate PCI devices, and even better, different cal-data lengths either entered into DTS or dynamically, and the MAC address patching as well.

**Comment by mpratt14** [comment]
I got successful flash from OEM (my silly mistake before)

I tested everything else with initramfs

here's my branch if you want to compare, but I ended up copying your partitioning Piotr
https://github.com/mpratt14/openwrt/commits/ath79-r6100

**Comment by pepe2k** [comment]
@adschm 
> I've never understood this properly. My uneducated impression from other people's comments was that it's _not_ same, but only one of them is right in a given situation. But this could be completely wrong either. I'd be happy if somebody could actually resolve this question, there is at least one other PR waiting for an answer as well.

As @mpratt14 wrote above, the `mtd-cal-data` is dedicated for SOC built-in WMACs, the user space scripts are used for external radios (PCI/e). I don't understand why other Netgear devices with SOC built-in radio use scripts instead of the DTS based approach. There are even boards using both approaches at the same time, e.g. 8devices Rambutan (see 326e488c7b02fb291e0b01970db3482284307c8d). I don't think that the `caldata_extract` is actually used in that case (`mtd-cal-data` will provide calibration data first). Could that be just a copy&paste mistake? I really don't think someone wanted to limit NAND reads by fetching the calibration data once and then keep it in filesystem? Check also this one: f5a9181d6e94044657df00d3531cf97f2e8d1d16

What's more, my R6100 doesn't even try to fetch calibration data using hotplug, even if `mtd-cal-data` is not included. Radio wakes up with some default `00:02:03:04:05:06` MAC and probably default calibration data stored on the SOC's OTP memory? I'm more and more convinced that something is totally messed up with use of `10-ath9k-eeprom`.

> It's probably easiest for most developers if GMAC and eth numbers align, i.e. &eth0 in DTS corresponds to eth0 at runtime. If I was asked, I'd set it up this way; I personally don't care much whether LAN/switch is eth0 or eth1 then. But having eth0 correspond to &eth1 is not really helpful....

I hope at some point DSA would make all of that issues go away.

If there are no more comments or issues, I would like to push version from my [staging tree](https://git.openwrt.org/openwrt/staging/pepe2k.git).
@mpratt14 let me know if it's ok to keep your SoB there.

**Comment by mpratt14** [comment]
@pepe2k like I said. WAN port should have the MAC located at 0x6 and internal switch should have MAC at 0x0

I believe the ethX numbers are the same in DTS and runtime correct?

everything else looks good 👍

**Comment by pepe2k** [comment]
@mpratt14 
> like I said. WAN port should have the MAC located at 0x6 and internal switch should have MAC at 0x0

Correct.

> I believe the ethX numbers are the same in DTS and runtime correct?

No, because of this:
```
[    0.907164] ag71xx 19000000.eth: Could not connect to PHY device. Deferring probe.
...
[    1.655600] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.665399] eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
```

So the `eth0: eth@19000000` from DTS becomes:
```
[    2.137960] ag71xx 19000000.eth: connected to PHY at mdio.0:1f:00 [uid=004dd042, driver=Generic PHY]
[    2.148222] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
```

That's how it works in ath79 and was discussed multiple times before.

> everything else looks good

OK!

Cheers,
Piotr

**Comment by mpratt14** [comment]
OK I see now. thanks for that summary

so OEM is aligned between gmacX and ethX, and in Openwrt it is reverse, but labeled in DTS as if it was not reverse...

**Comment by wigyori** [comment]
Hi,

> DTS only based version from @wigyori looks better, I updated my patch.

> * DTS: `firmware` partition: I removed it from my patch, we don't need it (maybe someone though it could be used for OpenWrt -> vendor firmware change)
> * DTS: `enable_gpio_11`: didn't need it on my device, the U-Boot already set up all the GPIOs for LEDs
> * power LED: I followed previous support and vendor's firmware (green for `booting` and `running`)
> * WAN LED: same as above, the vendor keeps amber WAN LED on when there is no Internet access (switches to green otherwise)
My take is still that given that we do not use the amber LED for anything, we could/should use it for indicating the LAN traffic. No hard feelings of course if everyone thinks we should not. :) 

> * WLAN LED: I use it for 2.4 GHz radio, @wigyori seems to be using it for 5 GHz, vendor uses this LED for both devices at the same time
> * DTS: `gpio-export` instead of `gpio-hog`: that's my personal preference which allows end user to (easily) control USB power
> * DTS: `chosen`: already defined in ar934x.dtsi
> * @adschm should we also add `DEVICE_COMPAT_VERSION` and `DEVICE_COMPAT_MESSAGE` for the kernel partition size change warning? Sysupgrade from <= 19.07 will fail anyway (the kernel partition size and supported devices don't match) but I haven't actually followed discussion about the COMPAT_VERSION and don't know the correct approach here.
That's a good point - I didn't think about upgrading from earlier versions.

> @wigyori any comments? Are you fine with adding your SoB to my version and push it upstream?
Just the above - "approved" if the SoB request still applies.

Thanks,
Zoltan H

**Pull request closed without merging**
